runtime-modules/storage/src/lib.rs

```diff
@@ -1,20 +1,2373 @@
+            ),
+        };
+
+        Voucher {
+            objects_used,
+            size_used,
+            ..voucher.clone()
+        }
+    }
+
+    // Adds a single object data to the voucher update (updates objects size and number).
+    fn add_object(&mut self, size: u64) {
+        self.objects_number = self.objects_number.saturating_add(1);
+        self.objects_total_size = self.objects_total_size.saturating_add(size);
+    }
+}
+
+/// Defines the storage bucket connection to the storage operator (storage WG worker).
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Debug)]
+pub enum StorageBucketOperatorStatus<WorkerId> {
+    /// No connection.
+    Missing,
+
+    /// Storage operator was invited.
+    InvitedStorageWorker(WorkerId),
+
+    /// Storage operator accepted the invitation.
+    StorageWorker(WorkerId),
+}
+
+impl<WorkerId> Default for StorageBucketOperatorStatus<WorkerId> {
+    fn default() -> Self {
+        Self::Missing
+    }
+}
+
+/// A commitment to hold some set of bags for long term storage. A bucket may have a bucket
+/// operator, which is a single worker in the storage working group.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Default, Clone, PartialEq, Eq, Debug)]
+pub struct StorageBucket<WorkerId> {
+    /// Current storage operator status.
+    pub operator_status: StorageBucketOperatorStatus<WorkerId>,
+
+    /// Defines whether the bucket accepts new bags.
+    pub accepting_new_bags: bool,
+
+    /// Defines limits for a bucket.
+    pub voucher: Voucher,
+
+    /// Defines storage bucket medata (like current storage provider URL).
+    pub metadata: Vec<u8>,
+}
+
+// Helper-struct for the data object uploading.
+#[derive(Default, Clone, Debug)]
+struct DataObjectCandidates<T: Trait> {
+    // next data object ID to be saved in the storage.
+    next_data_object_id: T::DataObjectId,
+
+    // 'ID-data object' map.
+    data_objects_map: BTreeMap<T::DataObjectId, DataObject<BalanceOf<T>>>,
+}
+
+/// Type alias for the DynamicBagObject.
+pub type DynamicBag<T> = DynamicBagObject<
+    <T as Trait>::DataObjectId,
+    <T as Trait>::StorageBucketId,
+    <T as Trait>::DistributionBucketId,
+    BalanceOf<T>,
+>;
+
+/// Dynamic bag container.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Default, Clone, PartialEq, Eq, Debug)]
+pub struct DynamicBagObject<
+    DataObjectId: Ord,
+    StorageBucketId: Ord,
+    DistributionBucketId: Ord,
+    Balance,
+> {
+    /// Associated data objects.
+    pub objects: BTreeMap<DataObjectId, DataObject<Balance>>,
+
+    /// Associated storage buckets.
+    pub stored_by: BTreeSet<StorageBucketId>,
+
+    /// Associated distribution buckets.
+    pub distributed_by: BTreeSet<DistributionBucketId>,
+
+    /// Dynamic bag deletion prize.
+    pub deletion_prize: Balance,
```

Comment:
It spares us from iteration through all data objects (could be expensive).

Reply:
1) Does that mean it is redundant, it is the sum of all the data objects?
2) When the objects are inlined, then its just iterating in memory, that is going to be super fast, and there is even going to be a length cap, so I don't think that is a real issue.
3) I believe we are still going for non-inlined dataobjects, at least last we spoke I argued pattionatley for this 🙏 📿 , in which case deletion of non-empty bags is not going to be a thing anyway.

Reply:
1. 'Redundant' in the meaning of 'cached' or 'denormalized'.
2. I'm not sure 'super fast' would be still the case for large bags (starting with hundreds of thousands data objects). However, a large bag deletion looks like a very rare operation. And we're committing a primordial sin of preliminary optimization. So, let's try to drop the field until stress tests show us the light of wisdom. There are additional implications - see another comment about the `BagManager`.
3. Please, remind me: didn't we agree to prevent deleting non-empty bags forcing the leader to manually delete possibly thousands of objects considering it as a rare operation, and drop the inlined data objects map in favor of a double map `BagId => DataObjectId => DataObject`. If we don't prevent non-empty bag deletion - we need to iterate through the double-map prefix by BagId. The only iteration would be also on deleting the dynamic bag.

Reply:
1. I reviewed this again, and I think the rationale for having this is distinct from the data objects. Remember, even if you have 0 videos in your channel, for example, we still need to give you an incentive to throw out the channel itself. That is what this prize does. This means it is needed, even if we have non-inlined objects. However, I do wonder, should we have the calling code manage this incentive itself, thereby dropping the distinction between dynamic and static bags in this module? In this alternative approach, a channel would have its own storage prize in the content directory, and that prize would be calibrated by that code to give sufficient incentive to cleanup both that channel there, and the dynamic bag in the storage system. Let me know what you think.
3. Yes I think this is what we agreed, this is what  I was trying to say. Note that the storage lead is not the deleter, it has to be the owner of the bag.